// @google-cloud/paginator/build/src/index.js
"use strict";
/*!
 * Copyright 2015 Google Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
/*!
 * @module common/paginator
 */
const arrify = require("arrify");
const extend = require("extend");
const resource_stream_1 = require("./resource-stream");
exports.ResourceStream = resource_stream_1.ResourceStream;
/*! Developer Documentation
 *
 * paginator is used to auto-paginate `nextQuery` methods as well as
 * streamifying them.
 *
 * Before:
 *
 *   search.query('done=true', function(err, results, nextQuery) {
 *     search.query(nextQuery, function(err, results, nextQuery) {***REMOVED***);
 * ***REMOVED***);
 *
 * After:
 *
 *   search.query('done=true', function(err, results) {***REMOVED***);
 *
 * Methods to extend should be written to accept callbacks and return a
 * `nextQuery`.
 */
class Paginator {
    /**
     * Cache the original method, then overwrite it on the Class's prototype.
     *
     * @param {function***REMOVED*** Class - The parent class of the methods to extend.
     * @param {string|string[]***REMOVED*** methodNames - Name(s) of the methods to extend.
     */
    // tslint:disable-next-line:variable-name
    extend(Class, methodNames) {
        methodNames = arrify(methodNames);
        methodNames.forEach(methodName => {
            const originalMethod = Class.prototype[methodName];
            // map the original method to a private member
            Class.prototype[methodName + '_'] = originalMethod;
            // overwrite the original to auto-paginate
            // tslint:disable-next-line:no-any
            Class.prototype[methodName] = function (...args) {
                const parsedArguments = paginator.parseArguments_(args);
                return paginator.run_(parsedArguments, originalMethod.bind(this));
          ***REMOVED***
      ***REMOVED***);
  ***REMOVED***
    /**
     * Wraps paginated API calls in a readable object stream.
     *
     * This method simply calls the nextQuery recursively, emitting results to a
     * stream. The stream ends when `nextQuery` is null.
     *
     * `maxResults` will act as a cap for how many results are fetched and emitted
     * to the stream.
     *
     * @param {string***REMOVED*** methodName - Name of the method to streamify.
     * @return {function***REMOVED*** - Wrapped function.
     */
    // tslint:disable-next-line:no-any
    streamify(methodName) {
        return function (
        // tslint:disable-next-line:no-any
        ...args) {
            const parsedArguments = paginator.parseArguments_(args);
            const originalMethod = this[methodName + '_'] || this[methodName];
            return paginator.runAsStream_(parsedArguments, originalMethod.bind(this));
      ***REMOVED***
  ***REMOVED***
    /**
     * Parse a pseudo-array `arguments` for a query and callback.
     *
     * @param {array***REMOVED*** args - The original `arguments` pseduo-array that the original
     *     method received.
     */
    // tslint:disable-next-line:no-any
    parseArguments_(args) {
        let query;
        let autoPaginate = true;
        let maxApiCalls = -1;
        let maxResults = -1;
        let callback;
        const firstArgument = args[0];
        const lastArgument = args[args.length - 1];
        if (typeof firstArgument === 'function') {
            callback = firstArgument;
      ***REMOVED***
        else {
            query = firstArgument;
      ***REMOVED***
        if (typeof lastArgument === 'function') {
            callback = lastArgument;
      ***REMOVED***
        if (typeof query === 'object') {
            query = extend(true, {***REMOVED***, query);
            // Check if the user only asked for a certain amount of results.
            if (query.maxResults && typeof query.maxResults === 'number') {
                // `maxResults` is used API-wide.
                maxResults = query.maxResults;
          ***REMOVED***
            else if (typeof query.pageSize === 'number') {
                // `pageSize` is Pub/Sub's `maxResults`.
                maxResults = query.pageSize;
          ***REMOVED***
            if (query.maxApiCalls && typeof query.maxApiCalls === 'number') {
                maxApiCalls = query.maxApiCalls;
                delete query.maxApiCalls;
          ***REMOVED***
            // maxResults is the user specified limit.
            if (maxResults !== -1 || query.autoPaginate === false) {
                autoPaginate = false;
          ***REMOVED***
      ***REMOVED***
        const parsedArguments = {
            query: query || {***REMOVED***,
            autoPaginate,
            maxApiCalls,
            maxResults,
            callback,
      ***REMOVED***
        parsedArguments.streamOptions = extend(true, {***REMOVED***, parsedArguments.query);
        delete parsedArguments.streamOptions.autoPaginate;
        delete parsedArguments.streamOptions.maxResults;
        delete parsedArguments.streamOptions.pageSize;
        return parsedArguments;
  ***REMOVED***
    /**
     * This simply checks to see if `autoPaginate` is set or not, if it's true
     * then we buffer all results, otherwise simply call the original method.
     *
     * @param {array***REMOVED*** parsedArguments - Parsed arguments from the original method
     *     call.
     * @param {object=|string=***REMOVED*** parsedArguments.query - Query object. This is most
     *     commonly an object, but to make the API more simple, it can also be a
     *     string in some places.
     * @param {function=***REMOVED*** parsedArguments.callback - Callback function.
     * @param {boolean***REMOVED*** parsedArguments.autoPaginate - Auto-pagination enabled.
     * @param {boolean***REMOVED*** parsedArguments.maxApiCalls - Maximum API calls to make.
     * @param {number***REMOVED*** parsedArguments.maxResults - Maximum results to return.
     * @param {function***REMOVED*** originalMethod - The cached method that accepts a callback
     *     and returns `nextQuery` to receive more results.
     */
    run_(parsedArguments, originalMethod) {
        const query = parsedArguments.query;
        const callback = parsedArguments.callback;
        if (!parsedArguments.autoPaginate) {
            return originalMethod(query, callback);
      ***REMOVED***
        const results = new Array();
        const promise = new Promise((resolve, reject) => {
            paginator
                .runAsStream_(parsedArguments, originalMethod)
                .on('error', reject)
                .on('data', (data) => results.push(data))
                .on('end', () => resolve(results));
      ***REMOVED***);
        if (!callback) {
            return promise.then(results => [results]);
      ***REMOVED***
        promise.then(results => callback(null, results), (err) => callback(err));
  ***REMOVED***
    /**
     * This method simply calls the nextQuery recursively, emitting results to a
     * stream. The stream ends when `nextQuery` is null.
     *
     * `maxResults` will act as a cap for how many results are fetched and emitted
     * to the stream.
     *
     * @param {object=|string=***REMOVED*** parsedArguments.query - Query object. This is most
     *     commonly an object, but to make the API more simple, it can also be a
     *     string in some places.
     * @param {function=***REMOVED*** parsedArguments.callback - Callback function.
     * @param {boolean***REMOVED*** parsedArguments.autoPaginate - Auto-pagination enabled.
     * @param {boolean***REMOVED*** parsedArguments.maxApiCalls - Maximum API calls to make.
     * @param {number***REMOVED*** parsedArguments.maxResults - Maximum results to return.
     * @param {function***REMOVED*** originalMethod - The cached method that accepts a callback
     *     and returns `nextQuery` to receive more results.
     * @return {stream***REMOVED*** - Readable object stream.
     */
    // tslint:disable-next-line:no-any
    runAsStream_(parsedArguments, originalMethod) {
        return new resource_stream_1.ResourceStream(parsedArguments, originalMethod);
  ***REMOVED***
***REMOVED***
exports.Paginator = Paginator;
const paginator = new Paginator();
exports.paginator = paginator;
