// @firebase/installations/src/util/errors.ts
/**
 * @license
 * Copyright 2019 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { ErrorFactory, FirebaseError ***REMOVED*** from '@firebase/util';
import { SERVICE, SERVICE_NAME ***REMOVED*** from './constants';

export const enum ErrorCode {
  MISSING_APP_CONFIG_VALUES = 'missing-app-config-values',
  CREATE_INSTALLATION_FAILED = 'create-installation-failed',
  GENERATE_TOKEN_FAILED = 'generate-token-failed',
  NOT_REGISTERED = 'not-registered',
  INSTALLATION_NOT_FOUND = 'installation-not-found',
  REQUEST_FAILED = 'request-failed',
  APP_OFFLINE = 'app-offline',
  DELETE_PENDING_REGISTRATION = 'delete-pending-registration'
***REMOVED***

const ERROR_DESCRIPTION_MAP: { readonly [key in ErrorCode]: string ***REMOVED*** = {
  [ErrorCode.MISSING_APP_CONFIG_VALUES]: 'Missing App configuration values.',
  [ErrorCode.CREATE_INSTALLATION_FAILED]:
    'Could not register Firebase Installation.',
  [ErrorCode.GENERATE_TOKEN_FAILED]: 'Could not generate Auth Token.',
  [ErrorCode.NOT_REGISTERED]: 'Firebase Installation is not registered.',
  [ErrorCode.INSTALLATION_NOT_FOUND]: 'Firebase Installation not found.',
  [ErrorCode.REQUEST_FAILED]:
    '{$requestName***REMOVED*** request failed with error "{$serverCode***REMOVED*** {$serverStatus***REMOVED***: {$serverMessage***REMOVED***"',
  [ErrorCode.APP_OFFLINE]: 'Could not process request. Application offline.',
  [ErrorCode.DELETE_PENDING_REGISTRATION]:
    "Can't delete installation while there is a pending registration request."
***REMOVED***;

interface ErrorParams {
  [ErrorCode.REQUEST_FAILED]: {
    requestName: string;
***REMOVED*** & ServerErrorData;
***REMOVED***

export const ERROR_FACTORY = new ErrorFactory<ErrorCode, ErrorParams>(
  SERVICE,
  SERVICE_NAME,
  ERROR_DESCRIPTION_MAP
);

export interface ServerErrorData {
  serverCode: number;
  serverMessage: string;
  serverStatus: string;
***REMOVED***

export type ServerError = FirebaseError & ServerErrorData;

/** Returns true if error is a FirebaseError that is based on an error from the server. */
export function isServerError(error: unknown): error is ServerError {
  return (
    error instanceof FirebaseError &&
    error.code.includes(ErrorCode.REQUEST_FAILED)
  );
***REMOVED***
