// needle/lib/parsers.js
//////////////////////////////////////////
// Defines mappings between content-type
// and the appropriate parsers.
//////////////////////////////////////////

var Transform = require('stream').Transform;
var sax = require('sax');

function parseXML(str, cb) {
  var obj, current, parser = sax.parser(true, { trim: true, lowercase: true ***REMOVED***)
  parser.onerror = parser.onend = done;

  function done(err) {
    parser.onerror = parser.onend = function() { ***REMOVED***
    cb(err, obj)
***REMOVED***

  function newElement(name, attributes) {
    return {
      name: name || '',
      value: '',
      attributes: attributes || {***REMOVED***,
      children: []
  ***REMOVED***
***REMOVED***

  parser.ontext = function(t) {
    if (current) current.value += t
***REMOVED***

  parser.onopentag = function(node) {
    var element = newElement(node.name, node.attributes)
    if (current) {
      element.parent = current
      current.children.push(element)
  ***REMOVED*** else { // root object
      obj = element
  ***REMOVED***

    current = element
***REMOVED***

  parser.onclosetag = function() {
    if (typeof current.parent !== 'undefined') {
      var just_closed = current
      current = current.parent
      delete just_closed.parent
  ***REMOVED***
***REMOVED***

  parser.write(str).close()
***REMOVED***

function parserFactory(name, fn) {

  function parser() {
    var chunks = [],
        stream = new Transform({ objectMode: true ***REMOVED***);

    // Buffer all our data
    stream._transform = function(chunk, encoding, done) {
      chunks.push(chunk);
      done();
  ***REMOVED***

    // And call the parser when all is there.
    stream._flush = function(done) {
      var self = this,
          data = Buffer.concat(chunks);

      try {
        fn(data, function(err, result) {
          if (err) throw err;
          self.push(result);
      ***REMOVED***);
    ***REMOVED*** catch (err) {
        self.push(data); // just pass the original data
    ***REMOVED*** finally {
        done();
    ***REMOVED***
  ***REMOVED***

    return stream;
***REMOVED***

  return { fn: parser, name: name ***REMOVED***;
***REMOVED***

var parsers = {***REMOVED***

function buildParser(name, types, fn) {
  var parser = parserFactory(name, fn);
  types.forEach(function(type) {
    parsers[type] = parser;
***REMOVED***)
***REMOVED***

buildParser('json', [
  'application/json',
  'text/javascript'
], function(buffer, cb) {
  var err, data;
  try { data = JSON.parse(buffer); ***REMOVED*** catch (e) { err = e; ***REMOVED***
  cb(err, data);
***REMOVED***);

buildParser('xml', [
  'text/xml',
  'application/xml',
  'application/rdf+xml',
  'application/rss+xml',
  'application/atom+xml'
], function(buffer, cb) {
  parseXML(buffer.toString(), function(err, obj) {
    cb(err, obj)
***REMOVED***)
***REMOVED***);

module.exports = parsers;
module.exports.use = buildParser;
