// date-and-time/date-and-time.js
/**
 * @preserve date-and-time.js (c) KNOWLEDGECODE | MIT
 */
(function (global) {
    'use strict';

    var date = {***REMOVED***,
        locales = {***REMOVED***,
        plugins = {***REMOVED***,
        lang = 'en',
        _res = {
            MMMM: ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'],
            MMM: ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'],
            dddd: ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'],
            ddd: ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'],
            dd: ['Su', 'Mo', 'Tu', 'We', 'Th', 'Fr', 'Sa'],
            A: ['a.m.', 'p.m.']
      ***REMOVED***,
        _formatter = {
            YYYY: function (d/*, formatString*/) { return ('000' + d.getFullYear()).slice(-4); ***REMOVED***,
            YY: function (d/*, formatString*/) { return ('0' + d.getFullYear()).slice(-2); ***REMOVED***,
            Y: function (d/*, formatString*/) { return '' + d.getFullYear(); ***REMOVED***,
            MMMM: function (d/*, formatString*/) { return this.res.MMMM[d.getMonth()]; ***REMOVED***,
            MMM: function (d/*, formatString*/) { return this.res.MMM[d.getMonth()]; ***REMOVED***,
            MM: function (d/*, formatString*/) { return ('0' + (d.getMonth() + 1)).slice(-2); ***REMOVED***,
            M: function (d/*, formatString*/) { return '' + (d.getMonth() + 1); ***REMOVED***,
            DD: function (d/*, formatString*/) { return ('0' + d.getDate()).slice(-2); ***REMOVED***,
            D: function (d/*, formatString*/) { return '' + d.getDate(); ***REMOVED***,
            HH: function (d/*, formatString*/) { return ('0' + d.getHours()).slice(-2); ***REMOVED***,
            H: function (d/*, formatString*/) { return '' + d.getHours(); ***REMOVED***,
            A: function (d/*, formatString*/) { return this.res.A[d.getHours() > 11 | 0]; ***REMOVED***,
            hh: function (d/*, formatString*/) { return ('0' + (d.getHours() % 12 || 12)).slice(-2); ***REMOVED***,
            h: function (d/*, formatString*/) { return '' + (d.getHours() % 12 || 12); ***REMOVED***,
            mm: function (d/*, formatString*/) { return ('0' + d.getMinutes()).slice(-2); ***REMOVED***,
            m: function (d/*, formatString*/) { return '' + d.getMinutes(); ***REMOVED***,
            ss: function (d/*, formatString*/) { return ('0' + d.getSeconds()).slice(-2); ***REMOVED***,
            s: function (d/*, formatString*/) { return '' + d.getSeconds(); ***REMOVED***,
            SSS: function (d/*, formatString*/) { return ('00' + d.getMilliseconds()).slice(-3); ***REMOVED***,
            SS: function (d/*, formatString*/) { return ('0' + (d.getMilliseconds() / 10 | 0)).slice(-2); ***REMOVED***,
            S: function (d/*, formatString*/) { return '' + (d.getMilliseconds() / 100 | 0); ***REMOVED***,
            dddd: function (d/*, formatString*/) { return this.res.dddd[d.getDay()]; ***REMOVED***,
            ddd: function (d/*, formatString*/) { return this.res.ddd[d.getDay()]; ***REMOVED***,
            dd: function (d/*, formatString*/) { return this.res.dd[d.getDay()]; ***REMOVED***,
            Z: function (d/*, formatString*/) {
                var offset = d.utc ? 0 : d.getTimezoneOffset() / 0.6;
                return (offset > 0 ? '-' : '+') + ('000' + Math.abs(offset - offset % 100 * 0.4)).slice(-4);
          ***REMOVED***,
            post: function (str) { return str; ***REMOVED***
      ***REMOVED***,
        _parser = {
            YYYY: function (str/*, formatString */) { return this.exec(/^\d{4***REMOVED***/, str); ***REMOVED***,
            YY: function (str/*, formatString */) {
                var result = this.exec(/^\d\d/, str);
                result.value += result.value < 70 ? 2000 : result.value < 100 ? 1900 : 0;
                return result;
          ***REMOVED***,
            Y: function (str/*, formatString */) { return this.exec(/^\d{1,4***REMOVED***/, str); ***REMOVED***,
            MMMM: function (str/*, formatString */) {
                var result = this.find(this.res.MMMM, str);
                result.value++;
                return result;
          ***REMOVED***,
            MMM: function (str/*, formatString */) {
                var result = this.find(this.res.MMM, str);
                result.value++;
                return result;
          ***REMOVED***,
            MM: function (str/*, formatString */) { return this.exec(/^\d\d/, str); ***REMOVED***,
            M: function (str/*, formatString */) { return this.exec(/^\d\d?/, str); ***REMOVED***,
            DD: function (str/*, formatString */) { return this.exec(/^\d\d/, str); ***REMOVED***,
            D: function (str/*, formatString */) { return this.exec(/^\d\d?/, str); ***REMOVED***,
            HH: function (str/*, formatString */) { return this.exec(/^\d\d/, str); ***REMOVED***,
            H: function (str/*, formatString */) { return this.exec(/^\d\d?/, str); ***REMOVED***,
            A: function (str/*, formatString */) { return this.find(this.res.A, str); ***REMOVED***,
            hh: function (str/*, formatString */) { return this.exec(/^\d\d/, str); ***REMOVED***,
            h: function (str/*, formatString */) { return this.exec(/^\d\d?/, str); ***REMOVED***,
            mm: function (str/*, formatString */) { return this.exec(/^\d\d/, str); ***REMOVED***,
            m: function (str/*, formatString */) { return this.exec(/^\d\d?/, str); ***REMOVED***,
            ss: function (str/*, formatString */) { return this.exec(/^\d\d/, str); ***REMOVED***,
            s: function (str/*, formatString */) { return this.exec(/^\d\d?/, str); ***REMOVED***,
            SSS: function (str/*, formatString */) { return this.exec(/^\d{1,3***REMOVED***/, str); ***REMOVED***,
            SS: function (str/*, formatString */) {
                var result = this.exec(/^\d\d?/, str);
                result.value *= 10;
                return result;
          ***REMOVED***,
            S: function (str/*, formatString */) {
                var result = this.exec(/^\d/, str);
                result.value *= 100;
                return result;
          ***REMOVED***,
            h12: function (h, a) { return (h === 12 ? 0 : h) + a * 12; ***REMOVED***,
            exec: function (re, str) {
                var result = (re.exec(str) || [''])[0];
                return { value: result | 0, length: result.length ***REMOVED***;
          ***REMOVED***,
            find: function (array, str) {
                var index = -1, length = 0;

                for (var i = 0, len = array.length, item; i < len; i++) {
                    item = array[i];
                    if (!str.indexOf(item) && item.length > length) {
                        index = i;
                        length = item.length;
                  ***REMOVED***
              ***REMOVED***
                return { value: index, length: length ***REMOVED***;
          ***REMOVED***,
            pre: function (str) { return str; ***REMOVED***
      ***REMOVED***,
        customize = function (code, base, locale) {
            var extend = function (proto, props, res) {
                    var Locale = function (r) {
                        if (r) { this.res = r; ***REMOVED***
                  ***REMOVED***

                    Locale.prototype = proto;
                    Locale.prototype.constructor = Locale;

                    var newLocale = new Locale(res),
                        value;

                    for (var key in props || {***REMOVED***) {
                        if (props.hasOwnProperty(key)) {
                            value = props[key];
                            newLocale[key] = value.slice ? value.slice() : value;
                      ***REMOVED***
                  ***REMOVED***
                    return newLocale;
              ***REMOVED***,
                loc = { res: extend(base.res, locale.res) ***REMOVED***;

            loc.formatter = extend(base.formatter, locale.formatter, loc.res);
            loc.parser = extend(base.parser, locale.parser, loc.res);
            locales[code] = loc;
      ***REMOVED***

    /**
     * formatting a date
     * @param {Date***REMOVED*** dateObj - a Date object
     * @param {string***REMOVED*** formatString - a format string
     * @param {boolean***REMOVED*** [utc] - output as UTC
     * @returns {string***REMOVED*** a formatted string
     */
    date.format = function (dateObj, formatString, utc) {
        var d = date.addMinutes(dateObj, utc ? dateObj.getTimezoneOffset() : 0),
            formatter = locales[lang].formatter;

        d.utc = utc;
        return formatString.replace(/\[[^\[\]]*]|\[.*\][^\[]*\]|([A-Za-z])\1*|./g, function (token) {
            return formatter[token] ? formatter.post(formatter[token](d, formatString)) : token.replace(/\[(.*)]/, '$1');
      ***REMOVED***);
  ***REMOVED***

    /**
     * pre-parsing a date string
     * @param {string***REMOVED*** dateString - a date string
     * @param {string***REMOVED*** formatString - a format string
     * @returns {Object***REMOVED*** a date structure
     */
    date.preparse = function (dateString, formatString) {
        var parser = locales[lang].parser,
            re = /([A-Za-z])\1*|./g,
            keys, token, result, offset = 0,
            dt = { Y: 1970, M: 1, D: 1, H: 0, A: 0, h: 0, m: 0, s: 0, S: 0, _index: 0, _length: 0, _match: 0 ***REMOVED***;

        dateString = parser.pre(dateString);
        formatString = formatString.replace(/\[[^\[\]]*]|\[.*\][^\[]*\]/g, function (str) {
            return str.replace(/./g, ' ').slice(2);
      ***REMOVED***);
        while ((keys = re.exec(formatString))) {
            token = keys[0];
            if (parser[token]) {
                result = parser[token](dateString.slice(offset), formatString);
                if (!result.length) {
                    break;
              ***REMOVED***
                offset += result.length;
                dt[token.charAt(0)] = result.value;
                dt._match++;
          ***REMOVED*** else if (token === dateString.charAt(offset) || token === ' ') {
                offset++;
          ***REMOVED*** else {
                break;
          ***REMOVED***
      ***REMOVED***
        dt.H = dt.H || parser.h12(dt.h, dt.A);
        dt._index = offset;
        dt._length = dateString.length;
        return dt;
  ***REMOVED***

    /**
     * validation
     * @param {Object|string***REMOVED*** arg - a date structure or a date string
     * @param {string***REMOVED*** [formatString] - a format string
     * @returns {boolean***REMOVED*** whether the date string is a valid date
     */
    date.isValid = function (arg, formatString) {
        var dt = typeof arg === 'string' ? date.preparse(arg, formatString) : arg,
            last = [31, 28 + date.isLeapYear(dt.Y) | 0, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31][dt.M - 1];

        return !(
            dt._index < 1 || dt._length < 1 || dt._index - dt._length || dt._match < 1 ||
            dt.Y < 1 || dt.Y > 9999 || dt.M < 1 || dt.M > 12 || dt.D < 1 || dt.D > last ||
            dt.H > 23 || dt.H < 0 || dt.m > 59 || dt.m < 0 || dt.s > 59 || dt.s < 0 || dt.S > 999 || dt.S < 0
        );
  ***REMOVED***

    /**
     * parsing a date string
     * @param {string***REMOVED*** dateString - a date string
     * @param {string***REMOVED*** formatString - a format string
     * @param {boolean***REMOVED*** [utc] - input as UTC
     * @returns {Date***REMOVED*** a constructed date
     */
    date.parse = function (dateString, formatString, utc) {
        var dt = date.preparse(dateString, formatString), dateObj;

        if (date.isValid(dt)) {
            dt.M -= dt.Y < 100 ? 22801 : 1; // 22801 = 1900 * 12 + 1
            if (utc) {
                dateObj = new Date(Date.UTC(dt.Y, dt.M, dt.D, dt.H, dt.m, dt.s, dt.S));
          ***REMOVED*** else {
                dateObj = new Date(dt.Y, dt.M, dt.D, dt.H, dt.m, dt.s, dt.S);
          ***REMOVED***
            return dateObj;
      ***REMOVED***
        return new Date(NaN);
  ***REMOVED***

    /**
     * adding years
     * @param {Date***REMOVED*** dateObj - a date object
     * @param {number***REMOVED*** years - number of years to add
     * @returns {Date***REMOVED*** a date after adding the value
     */
    date.addYears = function (dateObj, years) {
        return date.addMonths(dateObj, years * 12);
  ***REMOVED***

    /**
     * adding months
     * @param {Date***REMOVED*** dateObj - a date object
     * @param {number***REMOVED*** months - number of months to add
     * @returns {Date***REMOVED*** a date after adding the value
     */
    date.addMonths = function (dateObj, months) {
        var d = new Date(dateObj.getTime());

        d.setMonth(d.getMonth() + months);
        return d;
  ***REMOVED***

    /**
     * adding days
     * @param {Date***REMOVED*** dateObj - a date object
     * @param {number***REMOVED*** days - number of days to add
     * @returns {Date***REMOVED*** a date after adding the value
     */
    date.addDays = function (dateObj, days) {
        var d = new Date(dateObj.getTime());

        d.setDate(d.getDate() + days);
        return d;
  ***REMOVED***

    /**
     * adding hours
     * @param {Date***REMOVED*** dateObj - a date object
     * @param {number***REMOVED*** hours - number of hours to add
     * @returns {Date***REMOVED*** a date after adding the value
     */
    date.addHours = function (dateObj, hours) {
        return date.addMilliseconds(dateObj, hours * 3600000);
  ***REMOVED***

    /**
     * adding minutes
     * @param {Date***REMOVED*** dateObj - a date object
     * @param {number***REMOVED*** minutes - number of minutes to add
     * @returns {Date***REMOVED*** a date after adding the value
     */
    date.addMinutes = function (dateObj, minutes) {
        return date.addMilliseconds(dateObj, minutes * 60000);
  ***REMOVED***

    /**
     * adding seconds
     * @param {Date***REMOVED*** dateObj - a date object
     * @param {number***REMOVED*** seconds - number of seconds to add
     * @returns {Date***REMOVED*** a date after adding the value
     */
    date.addSeconds = function (dateObj, seconds) {
        return date.addMilliseconds(dateObj, seconds * 1000);
  ***REMOVED***

    /**
     * adding milliseconds
     * @param {Date***REMOVED*** dateObj - a date object
     * @param {number***REMOVED*** milliseconds - number of milliseconds to add
     * @returns {Date***REMOVED*** a date after adding the value
     */
    date.addMilliseconds = function (dateObj, milliseconds) {
        return new Date(dateObj.getTime() + milliseconds);
  ***REMOVED***

    /**
     * subtracting
     * @param {Date***REMOVED*** date1 - a Date object
     * @param {Date***REMOVED*** date2 - a Date object
     * @returns {Object***REMOVED*** a result object subtracting date2 from date1
     */
    date.subtract = function (date1, date2) {
        var delta = date1.getTime() - date2.getTime();

        return {
            toMilliseconds: function () {
                return delta;
          ***REMOVED***,
            toSeconds: function () {
                return delta / 1000 | 0;
          ***REMOVED***,
            toMinutes: function () {
                return delta / 60000 | 0;
          ***REMOVED***,
            toHours: function () {
                return delta / 3600000 | 0;
          ***REMOVED***,
            toDays: function () {
                return delta / 86400000 | 0;
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***

    /**
     * leap year
     * @param {number***REMOVED*** y - year
     * @returns {boolean***REMOVED*** whether the year is a leap year
     */
    date.isLeapYear = function (y) {
        return (!(y % 4) && !!(y % 100)) || !(y % 400);
  ***REMOVED***

    /**
     * comparison of two dates
     * @param {Date***REMOVED*** date1 - a Date object
     * @param {Date***REMOVED*** date2 - a Date object
     * @returns {boolean***REMOVED*** whether the dates are the same day (times are ignored)
     */
    date.isSameDay = function (date1, date2) {
        return date.format(date1, 'YYYYMMDD') === date.format(date2, 'YYYYMMDD');
  ***REMOVED***

    /**
     * change locale or setting a new locale definition
     * @param {string***REMOVED*** [code] - language code
     * @param {Object***REMOVED*** [locale] - locale definition
     * @returns {string***REMOVED*** current language code
     */
    date.locale = function (code, locale) {
        if (locale) {
            customize(code, { res: _res, formatter: _formatter, parser: _parser ***REMOVED***, locale);
      ***REMOVED*** else if (code) {
            lang = code;
      ***REMOVED***
        return lang;
  ***REMOVED***

    /**
     * locale extension
     * @param {Object***REMOVED*** extension - locale extension
     * @returns {void***REMOVED***
     */
    date.extend = function (extension) {
        customize(lang, locales[lang], extension);
  ***REMOVED***

    /**
     * plugin import or definition
     * @param {string***REMOVED*** name - plugin name
     * @param {Object***REMOVED*** [extension] - locale extension
     * @returns {void***REMOVED***
     */
    date.plugin = function (name, extension) {
        plugins[name] = plugins[name] || extension;
        if (!extension && plugins[name]) {
            date.extend(plugins[name]);
      ***REMOVED***
  ***REMOVED***

    // Create default locale (English)
    date.locale(lang, {***REMOVED***);

    if (typeof module === 'object' && typeof module.exports === 'object') {
        module.exports = date;
  ***REMOVED*** else if (typeof define === 'function' && define.amd) {
        define([], function () {
            return date;
      ***REMOVED***);
  ***REMOVED*** else {
        global.date = date;
  ***REMOVED***

***REMOVED***(this));
