// @google-cloud/common/build/src/service-object.js
"use strict";
/*!
 * Copyright 2015 Google Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
/*!
 * @module common/service-object
 */
const promisify_1 = require("@google-cloud/promisify");
const arrify = require("arrify");
const events_1 = require("events");
const extend = require("extend");
const util_1 = require("./util");
/**
 * ServiceObject is a base class, meant to be inherited from by a "service
 * object," like a BigQuery dataset or Storage bucket.
 *
 * Most of the time, these objects share common functionality; they can be
 * created or deleted, and you can get or set their metadata.
 *
 * By inheriting from this class, a service object will be extended with these
 * shared behaviors. Note that any method can be overridden when the service
 * object requires specific behavior.
 */
// tslint:disable-next-line no-any
class ServiceObject extends events_1.EventEmitter {
    /*
     * @constructor
     * @alias module:common/service-object
     *
     * @private
     *
     * @param {object***REMOVED*** config - Configuration object.
     * @param {string***REMOVED*** config.baseUrl - The base URL to make API requests to.
     * @param {string***REMOVED*** config.createMethod - The method which creates this object.
     * @param {string=***REMOVED*** config.id - The identifier of the object. For example, the
     *     name of a Storage bucket or Pub/Sub topic.
     * @param {object=***REMOVED*** config.methods - A map of each method name that should be inherited.
     * @param {object***REMOVED*** config.methods[].reqOpts - Default request options for this
     *     particular method. A common use case is when `setMetadata` requires a
     *     `PUT` method to override the default `PATCH`.
     * @param {object***REMOVED*** config.parent - The parent service instance. For example, an
     *     instance of Storage if the object is Bucket.
     */
    constructor(config) {
        super();
        this.metadata = {***REMOVED***;
        this.baseUrl = config.baseUrl;
        this.parent = config.parent; // Parent class.
        this.id = config.id; // Name or ID (e.g. dataset ID, bucket name, etc).
        this.createMethod = config.createMethod;
        this.methods = config.methods || {***REMOVED***;
        this.interceptors = [];
        this.Promise = this.parent ? this.parent.Promise : undefined;
        if (config.methods) {
            Object.getOwnPropertyNames(ServiceObject.prototype)
                .filter(methodName => {
                return (
                // All ServiceObjects need `request`.
                // clang-format off
                !/^request/.test(methodName) &&
                    // clang-format on
                    // The ServiceObject didn't redefine the method.
                    // tslint:disable-next-line no-any
                    this[methodName] ===
                        // tslint:disable-next-line no-any
                        ServiceObject.prototype[methodName] &&
                    // This method isn't wanted.
                    !config.methods[methodName]);
          ***REMOVED***)
                .forEach(methodName => {
                // tslint:disable-next-line no-any
                this[methodName] = undefined;
          ***REMOVED***);
      ***REMOVED***
  ***REMOVED***
    create(optionsOrCallback, callback) {
        const self = this;
        const args = [this.id];
        if (typeof optionsOrCallback === 'function') {
            callback = optionsOrCallback;
      ***REMOVED***
        if (typeof optionsOrCallback === 'object') {
            args.push(optionsOrCallback);
      ***REMOVED***
        // Wrap the callback to return *this* instance of the object, not the
        // newly-created one.
        // tslint: disable-next-line no-any
        function onCreate(...args) {
            const [err, instance] = args;
            if (!err) {
                self.metadata = instance.metadata;
                args[1] = self; // replace the created `instance` with this one.
          ***REMOVED***
            callback(...args);
      ***REMOVED***
        args.push(onCreate);
        this.createMethod.apply(null, args);
  ***REMOVED***
    delete(optionsOrCallback, cb) {
        const [options, callback] = util_1.util.maybeOptionsOrCallback(optionsOrCallback, cb);
        const methodConfig = (typeof this.methods.delete === 'object' && this.methods.delete) || {***REMOVED***;
        const reqOpts = extend(true, {
            method: 'DELETE',
            uri: '',
      ***REMOVED***, methodConfig.reqOpts, {
            qs: options,
      ***REMOVED***);
        // The `request` method may have been overridden to hold any special
        // behavior. Ensure we call the original `request` method.
        ServiceObject.prototype.request.call(this, reqOpts, callback);
  ***REMOVED***
    exists(optionsOrCallback, cb) {
        const [options, callback] = util_1.util.maybeOptionsOrCallback(optionsOrCallback, cb);
        this.get(options, err => {
            if (err) {
                if (err.code === 404) {
                    callback(null, false);
              ***REMOVED***
                else {
                    callback(err);
              ***REMOVED***
                return;
          ***REMOVED***
            callback(null, true);
      ***REMOVED***);
  ***REMOVED***
    get(optionsOrCallback, cb) {
        const self = this;
        const [opts, callback] = util_1.util.maybeOptionsOrCallback(optionsOrCallback, cb);
        const options = Object.assign({***REMOVED***, opts);
        const autoCreate = options.autoCreate && typeof this.create === 'function';
        delete options.autoCreate;
        function onCreate(err, instance, apiResponse) {
            if (err) {
                if (err.code === 409) {
                    self.get(options, callback);
                    return;
              ***REMOVED***
                callback(err, null, apiResponse);
                return;
          ***REMOVED***
            callback(null, instance, apiResponse);
      ***REMOVED***
        this.getMetadata(options, (err, metadata) => {
            if (err) {
                if (err.code === 404 && autoCreate) {
                    const args = [];
                    if (Object.keys(options).length > 0) {
                        args.push(options);
                  ***REMOVED***
                    args.push(onCreate);
                    self.create(...args);
                    return;
              ***REMOVED***
                callback(err, null, metadata);
                return;
          ***REMOVED***
            callback(null, self, metadata);
      ***REMOVED***);
  ***REMOVED***
    getMetadata(optionsOrCallback, cb) {
        const [options, callback] = util_1.util.maybeOptionsOrCallback(optionsOrCallback, cb);
        const methodConfig = (typeof this.methods.getMetadata === 'object' &&
            this.methods.getMetadata) ||
            {***REMOVED***;
        const reqOpts = extend(true, {
            uri: '',
      ***REMOVED***, methodConfig.reqOpts, {
            qs: options,
      ***REMOVED***);
        // The `request` method may have been overridden to hold any special
        // behavior. Ensure we call the original `request` method.
        ServiceObject.prototype.request.call(this, reqOpts, (err, body, res) => {
            this.metadata = body;
            callback(err, this.metadata, res);
      ***REMOVED***);
  ***REMOVED***
    setMetadata(metadata, optionsOrCallback, cb) {
        const [options, callback] = util_1.util.maybeOptionsOrCallback(optionsOrCallback, cb);
        const methodConfig = (typeof this.methods.setMetadata === 'object' &&
            this.methods.setMetadata) ||
            {***REMOVED***;
        const reqOpts = extend(true, {***REMOVED***, {
            method: 'PATCH',
            uri: '',
      ***REMOVED***, methodConfig.reqOpts, {
            json: metadata,
            qs: options,
      ***REMOVED***);
        // The `request` method may have been overridden to hold any special
        // behavior. Ensure we call the original `request` method.
        ServiceObject.prototype.request.call(this, reqOpts, (err, body, res) => {
            this.metadata = body;
            callback(err, this.metadata, res);
      ***REMOVED***);
  ***REMOVED***
    request_(reqOpts, callback) {
        reqOpts = extend(true, {***REMOVED***, reqOpts);
        const isAbsoluteUrl = reqOpts.uri.indexOf('http') === 0;
        const uriComponents = [this.baseUrl, this.id || '', reqOpts.uri];
        if (isAbsoluteUrl) {
            uriComponents.splice(0, uriComponents.indexOf(reqOpts.uri));
      ***REMOVED***
        reqOpts.uri = uriComponents
            .filter(x => x.trim()) // Limit to non-empty strings.
            .map(uriComponent => {
            const trimSlashesRegex = /^\/*|\/*$/g;
            return uriComponent.replace(trimSlashesRegex, '');
      ***REMOVED***)
            .join('/');
        const childInterceptors = arrify(reqOpts.interceptors_);
        const localInterceptors = [].slice.call(this.interceptors);
        reqOpts.interceptors_ = childInterceptors.concat(localInterceptors);
        if (reqOpts.shouldReturnStream) {
            return this.parent.requestStream(reqOpts);
      ***REMOVED***
        this.parent.request(reqOpts, callback);
  ***REMOVED***
    request(reqOpts, callback) {
        this.request_(reqOpts, callback);
  ***REMOVED***
    /**
     * Make an authenticated API request.
     *
     * @param {object***REMOVED*** reqOpts - Request options that are passed to `request`.
     * @param {string***REMOVED*** reqOpts.uri - A URI relative to the baseUrl.
     */
    requestStream(reqOpts) {
        const opts = extend(true, reqOpts, { shouldReturnStream: true ***REMOVED***);
        return this.request_(opts);
  ***REMOVED***
***REMOVED***
exports.ServiceObject = ServiceObject;
promisify_1.promisifyAll(ServiceObject);
