// protobufjs/src/common.js
"use strict";
module.exports = common;

var commonRe = /\/|\./;

/**
 * Provides common type definitions.
 * Can also be used to provide additional google types or your own custom types.
 * @param {string***REMOVED*** name Short name as in `google/protobuf/[name].proto` or full file name
 * @param {Object.<string,*>***REMOVED*** json JSON definition within `google.protobuf` if a short name, otherwise the file's root definition
 * @returns {undefined***REMOVED***
 * @property {INamespace***REMOVED*** google/protobuf/any.proto Any
 * @property {INamespace***REMOVED*** google/protobuf/duration.proto Duration
 * @property {INamespace***REMOVED*** google/protobuf/empty.proto Empty
 * @property {INamespace***REMOVED*** google/protobuf/field_mask.proto FieldMask
 * @property {INamespace***REMOVED*** google/protobuf/struct.proto Struct, Value, NullValue and ListValue
 * @property {INamespace***REMOVED*** google/protobuf/timestamp.proto Timestamp
 * @property {INamespace***REMOVED*** google/protobuf/wrappers.proto Wrappers
 * @example
 * // manually provides descriptor.proto (assumes google/protobuf/ namespace and .proto extension)
 * protobuf.common("descriptor", descriptorJson);
 *
 * // manually provides a custom definition (uses my.foo namespace)
 * protobuf.common("my/foo/bar.proto", myFooBarJson);
 */
function common(name, json) {
    if (!commonRe.test(name)) {
        name = "google/protobuf/" + name + ".proto";
        json = { nested: { google: { nested: { protobuf: { nested: json ***REMOVED*** ***REMOVED*** ***REMOVED*** ***REMOVED*** ***REMOVED***;
  ***REMOVED***
    common[name] = json;
***REMOVED***

// Not provided because of limited use (feel free to discuss or to provide yourself):
//
// google/protobuf/descriptor.proto
// google/protobuf/source_context.proto
// google/protobuf/type.proto
//
// Stripped and pre-parsed versions of these non-bundled files are instead available as part of
// the repository or package within the google/protobuf directory.

common("any", {

    /**
     * Properties of a google.protobuf.Any message.
     * @interface IAny
     * @type {Object***REMOVED***
     * @property {string***REMOVED*** [typeUrl]
     * @property {Uint8Array***REMOVED*** [bytes]
     * @memberof common
     */
    Any: {
        fields: {
            type_url: {
                type: "string",
                id: 1
          ***REMOVED***,
            value: {
                type: "bytes",
                id: 2
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***
***REMOVED***);

var timeType;

common("duration", {

    /**
     * Properties of a google.protobuf.Duration message.
     * @interface IDuration
     * @type {Object***REMOVED***
     * @property {number|Long***REMOVED*** [seconds]
     * @property {number***REMOVED*** [nanos]
     * @memberof common
     */
    Duration: timeType = {
        fields: {
            seconds: {
                type: "int64",
                id: 1
          ***REMOVED***,
            nanos: {
                type: "int32",
                id: 2
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***
***REMOVED***);

common("timestamp", {

    /**
     * Properties of a google.protobuf.Timestamp message.
     * @interface ITimestamp
     * @type {Object***REMOVED***
     * @property {number|Long***REMOVED*** [seconds]
     * @property {number***REMOVED*** [nanos]
     * @memberof common
     */
    Timestamp: timeType
***REMOVED***);

common("empty", {

    /**
     * Properties of a google.protobuf.Empty message.
     * @interface IEmpty
     * @memberof common
     */
    Empty: {
        fields: {***REMOVED***
  ***REMOVED***
***REMOVED***);

common("struct", {

    /**
     * Properties of a google.protobuf.Struct message.
     * @interface IStruct
     * @type {Object***REMOVED***
     * @property {Object.<string,IValue>***REMOVED*** [fields]
     * @memberof common
     */
    Struct: {
        fields: {
            fields: {
                keyType: "string",
                type: "Value",
                id: 1
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***,

    /**
     * Properties of a google.protobuf.Value message.
     * @interface IValue
     * @type {Object***REMOVED***
     * @property {string***REMOVED*** [kind]
     * @property {0***REMOVED*** [nullValue]
     * @property {number***REMOVED*** [numberValue]
     * @property {string***REMOVED*** [stringValue]
     * @property {boolean***REMOVED*** [boolValue]
     * @property {IStruct***REMOVED*** [structValue]
     * @property {IListValue***REMOVED*** [listValue]
     * @memberof common
     */
    Value: {
        oneofs: {
            kind: {
                oneof: [
                    "nullValue",
                    "numberValue",
                    "stringValue",
                    "boolValue",
                    "structValue",
                    "listValue"
                ]
          ***REMOVED***
      ***REMOVED***,
        fields: {
            nullValue: {
                type: "NullValue",
                id: 1
          ***REMOVED***,
            numberValue: {
                type: "double",
                id: 2
          ***REMOVED***,
            stringValue: {
                type: "string",
                id: 3
          ***REMOVED***,
            boolValue: {
                type: "bool",
                id: 4
          ***REMOVED***,
            structValue: {
                type: "Struct",
                id: 5
          ***REMOVED***,
            listValue: {
                type: "ListValue",
                id: 6
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***,

    NullValue: {
        values: {
            NULL_VALUE: 0
      ***REMOVED***
  ***REMOVED***,

    /**
     * Properties of a google.protobuf.ListValue message.
     * @interface IListValue
     * @type {Object***REMOVED***
     * @property {Array.<IValue>***REMOVED*** [values]
     * @memberof common
     */
    ListValue: {
        fields: {
            values: {
                rule: "repeated",
                type: "Value",
                id: 1
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***
***REMOVED***);

common("wrappers", {

    /**
     * Properties of a google.protobuf.DoubleValue message.
     * @interface IDoubleValue
     * @type {Object***REMOVED***
     * @property {number***REMOVED*** [value]
     * @memberof common
     */
    DoubleValue: {
        fields: {
            value: {
                type: "double",
                id: 1
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***,

    /**
     * Properties of a google.protobuf.FloatValue message.
     * @interface IFloatValue
     * @type {Object***REMOVED***
     * @property {number***REMOVED*** [value]
     * @memberof common
     */
    FloatValue: {
        fields: {
            value: {
                type: "float",
                id: 1
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***,

    /**
     * Properties of a google.protobuf.Int64Value message.
     * @interface IInt64Value
     * @type {Object***REMOVED***
     * @property {number|Long***REMOVED*** [value]
     * @memberof common
     */
    Int64Value: {
        fields: {
            value: {
                type: "int64",
                id: 1
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***,

    /**
     * Properties of a google.protobuf.UInt64Value message.
     * @interface IUInt64Value
     * @type {Object***REMOVED***
     * @property {number|Long***REMOVED*** [value]
     * @memberof common
     */
    UInt64Value: {
        fields: {
            value: {
                type: "uint64",
                id: 1
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***,

    /**
     * Properties of a google.protobuf.Int32Value message.
     * @interface IInt32Value
     * @type {Object***REMOVED***
     * @property {number***REMOVED*** [value]
     * @memberof common
     */
    Int32Value: {
        fields: {
            value: {
                type: "int32",
                id: 1
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***,

    /**
     * Properties of a google.protobuf.UInt32Value message.
     * @interface IUInt32Value
     * @type {Object***REMOVED***
     * @property {number***REMOVED*** [value]
     * @memberof common
     */
    UInt32Value: {
        fields: {
            value: {
                type: "uint32",
                id: 1
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***,

    /**
     * Properties of a google.protobuf.BoolValue message.
     * @interface IBoolValue
     * @type {Object***REMOVED***
     * @property {boolean***REMOVED*** [value]
     * @memberof common
     */
    BoolValue: {
        fields: {
            value: {
                type: "bool",
                id: 1
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***,

    /**
     * Properties of a google.protobuf.StringValue message.
     * @interface IStringValue
     * @type {Object***REMOVED***
     * @property {string***REMOVED*** [value]
     * @memberof common
     */
    StringValue: {
        fields: {
            value: {
                type: "string",
                id: 1
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***,

    /**
     * Properties of a google.protobuf.BytesValue message.
     * @interface IBytesValue
     * @type {Object***REMOVED***
     * @property {Uint8Array***REMOVED*** [value]
     * @memberof common
     */
    BytesValue: {
        fields: {
            value: {
                type: "bytes",
                id: 1
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***
***REMOVED***);

common("field_mask", {

    /**
     * Properties of a google.protobuf.FieldMask message.
     * @interface IDoubleValue
     * @type {Object***REMOVED***
     * @property {number***REMOVED*** [value]
     * @memberof common
     */
    FieldMask: {
        fields: {
            paths: {
                rule: "repeated",
                type: "string",
                id: 1
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***
***REMOVED***);

/**
 * Gets the root definition of the specified common proto file.
 *
 * Bundled definitions are:
 * - google/protobuf/any.proto
 * - google/protobuf/duration.proto
 * - google/protobuf/empty.proto
 * - google/protobuf/field_mask.proto
 * - google/protobuf/struct.proto
 * - google/protobuf/timestamp.proto
 * - google/protobuf/wrappers.proto
 *
 * @param {string***REMOVED*** file Proto file name
 * @returns {INamespace|null***REMOVED*** Root definition or `null` if not defined
 */
common.get = function get(file) {
    return common[file] || null;
***REMOVED***;
