// grpc/src/common.js
/**
 * @license
 * Copyright 2015 gRPC authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

'use strict';

var constants = require('./constants');

/**
 * Wrap a function to pass null-like values through without calling it. If no
 * function is given, just uses the identity.
 * @private
 * @param {?function***REMOVED*** func The function to wrap
 * @return {function***REMOVED*** The wrapped function
 */
exports.wrapIgnoreNull = function wrapIgnoreNull(func) {
  if (!func) {
    return x => x;
***REMOVED***
  return function(arg) {
    if (arg === null || arg === undefined) {
      return null;
  ***REMOVED***
    return func(arg);
***REMOVED***
***REMOVED***;

/**
 * The logger object for the gRPC module. Defaults to console.
 * @private
 */
exports.logger = console;

/**
 * The current logging verbosity. 0 corresponds to logging everything
 * @private
 */
exports.logVerbosity = 0;

/**
 * Log a message if the severity is at least as high as the current verbosity
 * @private
 * @param {Number***REMOVED*** severity A value of the grpc.logVerbosity map
 * @param {String***REMOVED*** message The message to log
 */
exports.log = function log(severity, message) {
  if (severity >= exports.logVerbosity) {
    exports.logger.error(message);
***REMOVED***
***REMOVED***;

/**
 * Default options for loading proto files into gRPC
 * @alias grpc~defaultLoadOptions
 */
exports.defaultGrpcOptions = {
  convertFieldsToCamelCase: false,
  binaryAsBase64: false,
  longsAsStrings: true,
  enumsAsStrings: true,
  deprecatedArgumentOrder: false
***REMOVED***;

/**
 * Create an Error object from a status object
 * @param {grpc~StatusObject***REMOVED*** status The status object
 * @return {Error***REMOVED*** The resulting Error
 */
exports.createStatusError = function(status) {
  let inverted = Object.keys(constants.status)
    .reduce((acc, key) => {
      acc[constants.status[key]] = key;
      return acc;
  ***REMOVED***, {***REMOVED***);
  let statusName = inverted[status.code];
  let message = `${status.code***REMOVED*** ${statusName***REMOVED***: ${status.details***REMOVED***`;
  let error = new Error(message);
  error.code = status.code;
  error.metadata = status.metadata;
  error.details = status.details;
  return error;
***REMOVED***;

/**
 * Get a method's type from its definition
 * @param {grpc~MethodDefinition***REMOVED*** method_definition
 * @return {number***REMOVED***
 */
exports.getMethodType = function(method_definition) {
  if (method_definition.requestStream) {
    if (method_definition.responseStream) {
      return constants.methodTypes.BIDI_STREAMING;
  ***REMOVED*** else {
      return constants.methodTypes.CLIENT_STREAMING;
  ***REMOVED***
***REMOVED*** else {
    if (method_definition.responseStream) {
      return constants.methodTypes.SERVER_STREAMING;
  ***REMOVED*** else {
      return constants.methodTypes.UNARY;
  ***REMOVED***
***REMOVED***
***REMOVED***;

/**
 * Iterate over a collection of items, and run the given handler.
 * Return the results as a flattened array of values.
 *
 * @private
 *
 * @param {Array***REMOVED*** collection Array of items to process
 * @param {Function***REMOVED*** handler The function to call on each element in the array
 * @return {Array***REMOVED*** A flattened array of results.
 */
exports.flatMap = function(collection, handler) {
  const mapped = collection.map(handler);
  return mapped.reduce((acc, curr) => acc.concat(curr), []);
***REMOVED***

/**
 * Given an array of property names and an array of values,
 * combine the two into an object map.
 * Equivalent to _.zipObject.
 *
 * @private
 *
 * @param props {Array<String>***REMOVED*** Array of property names
 * @param values {Array***REMOVED*** Array of property values
 * @return {Object***REMOVED*** An object with the combined values
 */
exports.zipObject = function(props, values) {
  return props.reduce((acc, curr, idx) => {
    return Object.assign(acc, { [curr]: values[idx] ***REMOVED***);
***REMOVED***, {***REMOVED***);
***REMOVED***

// JSDoc definitions that are used in multiple other modules

/**
 * Represents the status of a completed request. If `code` is
 * {@link grpc.status***REMOVED***.OK, then the request has completed successfully.
 * Otherwise, the request has failed, `details` will contain a description of
 * the error. Either way, `metadata` contains the trailing response metadata
 * sent by the server when it finishes processing the call.
 * @typedef {object***REMOVED*** grpc~StatusObject
 * @property {number***REMOVED*** code The error code, a key of {@link grpc.status***REMOVED***
 * @property {string***REMOVED*** details Human-readable description of the status
 * @property {grpc.Metadata***REMOVED*** metadata Trailing metadata sent with the status,
 *     if applicable
 */

/**
 * Describes how a request has failed. The member `message` will be the same as
 * `details` in {@link grpc~StatusObject***REMOVED***, and `code` and `metadata` are the
 * same as in that object.
 * @typedef {Error***REMOVED*** grpc~ServiceError
 * @property {number***REMOVED*** code The error code, a key of {@link grpc.status***REMOVED*** that is
 *     not `grpc.status.OK`
 * @property {grpc.Metadata***REMOVED*** metadata Trailing metadata sent with the status,
 *     if applicable
 */

/**
 * The EventEmitter class in the event standard module
 * @external EventEmitter
 * @see https://nodejs.org/api/events.html#events_class_eventemitter
 */

/**
 * The Readable class in the stream standard module
 * @external Readable
 * @see https://nodejs.org/api/stream.html#stream_readable_streams
 */

/**
 * The Writable class in the stream standard module
 * @external Writable
 * @see https://nodejs.org/api/stream.html#stream_writable_streams
 */

/**
 * The Duplex class in the stream standard module
 * @external Duplex
 * @see https://nodejs.org/api/stream.html#stream_class_stream_duplex
 */

/**
 * A serialization function
 * @callback grpc~serialize
 * @param {****REMOVED*** value The value to serialize
 * @return {Buffer***REMOVED*** The value serialized as a byte sequence
 */

/**
 * A deserialization function
 * @callback grpc~deserialize
 * @param {Buffer***REMOVED*** data The byte sequence to deserialize
 * @return {****REMOVED*** The data deserialized as a value
 */

/**
 * The deadline of an operation. If it is a date, the deadline is reached at
 * the date and time specified. If it is a finite number, it is treated as
 * a number of milliseconds since the Unix Epoch. If it is Infinity, the
 * deadline will never be reached. If it is -Infinity, the deadline has already
 * passed.
 * @typedef {(number|Date)***REMOVED*** grpc~Deadline
 */

/**
 * An object that completely defines a service method signature.
 * @typedef {Object***REMOVED*** grpc~MethodDefinition
 * @property {string***REMOVED*** path The method's URL path
 * @property {boolean***REMOVED*** requestStream Indicates whether the method accepts
 *     a stream of requests
 * @property {boolean***REMOVED*** responseStream Indicates whether the method returns
 *     a stream of responses
 * @property {grpc~serialize***REMOVED*** requestSerialize Serialization
 *     function for request values
 * @property {grpc~serialize***REMOVED*** responseSerialize Serialization
 *     function for response values
 * @property {grpc~deserialize***REMOVED*** requestDeserialize Deserialization
 *     function for request data
 * @property {grpc~deserialize***REMOVED*** responseDeserialize Deserialization
 *     function for repsonse data
 */

/**
 * @function MetadataListener
 * @param {grpc.Metadata***REMOVED*** metadata The response metadata.
 * @param {function***REMOVED*** next Passes metadata to the next interceptor.
 */

/**
 * @function MessageListener
 * @param {jspb.Message***REMOVED*** message The response message.
 * @param {function***REMOVED*** next Passes a message to the next interceptor.
 */

/**
 * @function StatusListener
 * @param {grpc~StatusObject***REMOVED*** status The response status.
 * @param {function***REMOVED*** next Passes a status to the next interceptor.
 */

/**
 * A set of interceptor functions triggered by responses
 * @typedef {object***REMOVED*** grpc~Listener
 * @property {MetadataListener=***REMOVED*** onReceiveMetadata A function triggered by
 *     response metadata.
 * @property {MessageListener=***REMOVED*** onReceiveMessage A function triggered by a
 *     response message.
 * @property {StatusListener=***REMOVED*** onReceiveStatus A function triggered by a
 *     response status.
 */

/**
 * @function MetadataRequester
 * @param {grpc.Metadata***REMOVED*** metadata The request metadata.
 * @param {grpc~Listener***REMOVED*** listener A listener wired to the previous layers
 *     in the interceptor stack.
 * @param {function***REMOVED*** next Passes metadata and a listener to the next
 *      interceptor.
 */

/**
 * @function MessageRequester
 * @param {jspb.Message***REMOVED*** message The request message.
 * @param {function***REMOVED*** next Passes a message to the next interceptor.
 */

/**
 * @function CloseRequester
 * @param {function***REMOVED*** next Calls the next interceptor.
 */

/**
 * @function CancelRequester
 * @param {function***REMOVED*** next Calls the next interceptor.
 */

/**
 * @function GetPeerRequester
 * @param {function***REMOVED*** next Calls the next interceptor.
 * @return {string***REMOVED***
 */

/**
 * @typedef {object***REMOVED*** grpc~Requester
 * @param {MetadataRequester=***REMOVED*** start A function triggered when the call begins.
 * @param {MessageRequester=***REMOVED*** sendMessage A function triggered by the request
 *     message.
 * @param {CloseRequester=***REMOVED*** halfClose A function triggered when the client
 *     closes the call.
 * @param {CancelRequester=***REMOVED*** cancel A function triggered when the call is
 *     cancelled.
 * @param {GetPeerRequester=***REMOVED*** getPeer A function triggered when the endpoint is
 *     requested.
 */

/**
 * An object that completely defines a service.
 * @typedef {Object.<string, grpc~MethodDefinition>***REMOVED*** grpc~ServiceDefinition
 */

 /**
  * An object that defines a protobuf type
  * @typedef {object***REMOVED*** grpc~ProtobufTypeDefinition
  * @param {string***REMOVED*** format The format of the type definition object
  * @param {****REMOVED*** type The type definition object
  * @param {Buffer[]***REMOVED*** fileDescriptorProtos Binary protobuf file
  *     descriptors for all files loaded to construct this type
  */

/**
 * An object that defines a package hierarchy with multiple services
 * @typedef {Object.<string, grpc~ServiceDefinition|grpc~ProtobufTypeDefinition>***REMOVED*** grpc~PackageDefinition
 */

/**
 * A function for dynamically assigning an interceptor to a call.
 * @function InterceptorProvider
 * @param {grpc~MethodDefinition***REMOVED*** method_definition The method to provide
 *     an interceptor for.
 * @return {Interceptor|null***REMOVED*** The interceptor to provide or nothing
 */

/**
 * A function which can modify call options and produce methods to intercept
 * RPC operations.
 * @function Interceptor
 * @param {object***REMOVED*** options The grpc call options
 * @param {NextCall***REMOVED*** nextCall
 * @return {InterceptingCall***REMOVED***
 */

/**
 * A function which produces the next InterceptingCall.
 * @function NextCall
 * @param {object***REMOVED*** options The grpc call options
 * @return {InterceptingCall|null***REMOVED***
 */
