// date-and-time/plugin/meridiem.js
(function (global) {
    'use strict';

    var exec = function (date) {
        date.plugin('meridiem', {
            res: {
                A: ['a.m.', 'p.m.', 'AM', 'PM', 'A.M.', 'P.M.', 'am', 'pm']
          ***REMOVED***,
            formatter: {
                AA: function (d) {
                    // AM / PM
                    return this.res.A[d.getHours() > 11 | 0 + 2];
              ***REMOVED***,
                a: function (d) {
                    // A.M. / P.M.
                    return this.res.A[d.getHours() > 11 | 0 + 4];
              ***REMOVED***,
                aa: function (d) {
                    // am / pm
                    return this.res.A[d.getHours() > 11 | 0 + 6];
              ***REMOVED***
          ***REMOVED***,
            parser: {
                A: function (str) {
                    var result = this.find(this.res.A, str);
                    result.value %= 2;
                    return result;
              ***REMOVED***
          ***REMOVED***
      ***REMOVED***);
  ***REMOVED***

    if (typeof module === 'object' && typeof module.exports === 'object') {
        (module.paths || []).push('./');
        exec(require('date-and-time'));
  ***REMOVED*** else if (typeof define === 'function' && define.amd) {
        define(['date-and-time'], exec);
  ***REMOVED*** else {
        exec(global.date);
  ***REMOVED***

***REMOVED***(this));
