// promise-polyfill/lib/index.js
'use strict';

/**
 * @this {Promise***REMOVED***
 */
function finallyConstructor(callback) {
  var constructor = this.constructor;
  return this.then(
    function(value) {
      // @ts-ignore
      return constructor.resolve(callback()).then(function() {
        return value;
    ***REMOVED***);
  ***REMOVED***,
    function(reason) {
      // @ts-ignore
      return constructor.resolve(callback()).then(function() {
        // @ts-ignore
        return constructor.reject(reason);
    ***REMOVED***);
  ***REMOVED***
  );
***REMOVED***

// Store setTimeout reference so promise-polyfill will be unaffected by
// other code modifying setTimeout (like sinon.useFakeTimers())
var setTimeoutFunc = setTimeout;

function isArray(x) {
  return Boolean(x && typeof x.length !== 'undefined');
***REMOVED***

function noop() {***REMOVED***

// Polyfill for Function.prototype.bind
function bind(fn, thisArg) {
  return function() {
    fn.apply(thisArg, arguments);
***REMOVED***
***REMOVED***

/**
 * @constructor
 * @param {Function***REMOVED*** fn
 */
function Promise(fn) {
  if (!(this instanceof Promise))
    throw new TypeError('Promises must be constructed via new');
  if (typeof fn !== 'function') throw new TypeError('not a function');
  /** @type {!number***REMOVED*** */
  this._state = 0;
  /** @type {!boolean***REMOVED*** */
  this._handled = false;
  /** @type {Promise|undefined***REMOVED*** */
  this._value = undefined;
  /** @type {!Array<!Function>***REMOVED*** */
  this._deferreds = [];

  doResolve(fn, this);
***REMOVED***

function handle(self, deferred) {
  while (self._state === 3) {
    self = self._value;
***REMOVED***
  if (self._state === 0) {
    self._deferreds.push(deferred);
    return;
***REMOVED***
  self._handled = true;
  Promise._immediateFn(function() {
    var cb = self._state === 1 ? deferred.onFulfilled : deferred.onRejected;
    if (cb === null) {
      (self._state === 1 ? resolve : reject)(deferred.promise, self._value);
      return;
  ***REMOVED***
    var ret;
    try {
      ret = cb(self._value);
  ***REMOVED*** catch (e) {
      reject(deferred.promise, e);
      return;
  ***REMOVED***
    resolve(deferred.promise, ret);
***REMOVED***);
***REMOVED***

function resolve(self, newValue) {
  try {
    // Promise Resolution Procedure: https://github.com/promises-aplus/promises-spec#the-promise-resolution-procedure
    if (newValue === self)
      throw new TypeError('A promise cannot be resolved with itself.');
    if (
      newValue &&
      (typeof newValue === 'object' || typeof newValue === 'function')
    ) {
      var then = newValue.then;
      if (newValue instanceof Promise) {
        self._state = 3;
        self._value = newValue;
        finale(self);
        return;
    ***REMOVED*** else if (typeof then === 'function') {
        doResolve(bind(then, newValue), self);
        return;
    ***REMOVED***
  ***REMOVED***
    self._state = 1;
    self._value = newValue;
    finale(self);
***REMOVED*** catch (e) {
    reject(self, e);
***REMOVED***
***REMOVED***

function reject(self, newValue) {
  self._state = 2;
  self._value = newValue;
  finale(self);
***REMOVED***

function finale(self) {
  if (self._state === 2 && self._deferreds.length === 0) {
    Promise._immediateFn(function() {
      if (!self._handled) {
        Promise._unhandledRejectionFn(self._value);
    ***REMOVED***
  ***REMOVED***);
***REMOVED***

  for (var i = 0, len = self._deferreds.length; i < len; i++) {
    handle(self, self._deferreds[i]);
***REMOVED***
  self._deferreds = null;
***REMOVED***

/**
 * @constructor
 */
function Handler(onFulfilled, onRejected, promise) {
  this.onFulfilled = typeof onFulfilled === 'function' ? onFulfilled : null;
  this.onRejected = typeof onRejected === 'function' ? onRejected : null;
  this.promise = promise;
***REMOVED***

/**
 * Take a potentially misbehaving resolver function and make sure
 * onFulfilled and onRejected are only called once.
 *
 * Makes no guarantees about asynchrony.
 */
function doResolve(fn, self) {
  var done = false;
  try {
    fn(
      function(value) {
        if (done) return;
        done = true;
        resolve(self, value);
    ***REMOVED***,
      function(reason) {
        if (done) return;
        done = true;
        reject(self, reason);
    ***REMOVED***
    );
***REMOVED*** catch (ex) {
    if (done) return;
    done = true;
    reject(self, ex);
***REMOVED***
***REMOVED***

Promise.prototype['catch'] = function(onRejected) {
  return this.then(null, onRejected);
***REMOVED***;

Promise.prototype.then = function(onFulfilled, onRejected) {
  // @ts-ignore
  var prom = new this.constructor(noop);

  handle(this, new Handler(onFulfilled, onRejected, prom));
  return prom;
***REMOVED***;

Promise.prototype['finally'] = finallyConstructor;

Promise.all = function(arr) {
  return new Promise(function(resolve, reject) {
    if (!isArray(arr)) {
      return reject(new TypeError('Promise.all accepts an array'));
  ***REMOVED***

    var args = Array.prototype.slice.call(arr);
    if (args.length === 0) return resolve([]);
    var remaining = args.length;

    function res(i, val) {
      try {
        if (val && (typeof val === 'object' || typeof val === 'function')) {
          var then = val.then;
          if (typeof then === 'function') {
            then.call(
              val,
              function(val) {
                res(i, val);
            ***REMOVED***,
              reject
            );
            return;
        ***REMOVED***
      ***REMOVED***
        args[i] = val;
        if (--remaining === 0) {
          resolve(args);
      ***REMOVED***
    ***REMOVED*** catch (ex) {
        reject(ex);
    ***REMOVED***
  ***REMOVED***

    for (var i = 0; i < args.length; i++) {
      res(i, args[i]);
  ***REMOVED***
***REMOVED***);
***REMOVED***;

Promise.resolve = function(value) {
  if (value && typeof value === 'object' && value.constructor === Promise) {
    return value;
***REMOVED***

  return new Promise(function(resolve) {
    resolve(value);
***REMOVED***);
***REMOVED***;

Promise.reject = function(value) {
  return new Promise(function(resolve, reject) {
    reject(value);
***REMOVED***);
***REMOVED***;

Promise.race = function(arr) {
  return new Promise(function(resolve, reject) {
    if (!isArray(arr)) {
      return reject(new TypeError('Promise.race accepts an array'));
  ***REMOVED***

    for (var i = 0, len = arr.length; i < len; i++) {
      Promise.resolve(arr[i]).then(resolve, reject);
  ***REMOVED***
***REMOVED***);
***REMOVED***;

// Use polyfill for setImmediate for performance gains
Promise._immediateFn =
  // @ts-ignore
  (typeof setImmediate === 'function' &&
    function(fn) {
      // @ts-ignore
      setImmediate(fn);
  ***REMOVED***) ||
  function(fn) {
    setTimeoutFunc(fn, 0);
***REMOVED***

Promise._unhandledRejectionFn = function _unhandledRejectionFn(err) {
  if (typeof console !== 'undefined' && console) {
    console.warn('Possible Unhandled Promise Rejection:', err); // eslint-disable-line no-console
***REMOVED***
***REMOVED***;

module.exports = Promise;
