// array-flatten/array-flatten.js
'use strict'

/**
 * Expose `arrayFlatten`.
 */
module.exports = flatten
module.exports.from = flattenFrom
module.exports.depth = flattenDepth
module.exports.fromDepth = flattenFromDepth

/**
 * Flatten an array.
 *
 * @param  {Array***REMOVED*** array
 * @return {Array***REMOVED***
 */
function flatten (array) {
  if (!Array.isArray(array)) {
    throw new TypeError('Expected value to be an array')
***REMOVED***

  return flattenFrom(array)
***REMOVED***

/**
 * Flatten an array-like structure.
 *
 * @param  {Array***REMOVED*** array
 * @return {Array***REMOVED***
 */
function flattenFrom (array) {
  return flattenDown(array, [])
***REMOVED***

/**
 * Flatten an array-like structure with depth.
 *
 * @param  {Array***REMOVED***  array
 * @param  {number***REMOVED*** depth
 * @return {Array***REMOVED***
 */
function flattenDepth (array, depth) {
  if (!Array.isArray(array)) {
    throw new TypeError('Expected value to be an array')
***REMOVED***

  return flattenFromDepth(array, depth)
***REMOVED***

/**
 * Flatten an array-like structure with depth.
 *
 * @param  {Array***REMOVED***  array
 * @param  {number***REMOVED*** depth
 * @return {Array***REMOVED***
 */
function flattenFromDepth (array, depth) {
  if (typeof depth !== 'number') {
    throw new TypeError('Expected the depth to be a number')
***REMOVED***

  return flattenDownDepth(array, [], depth)
***REMOVED***

/**
 * Flatten an array indefinitely.
 *
 * @param  {Array***REMOVED*** array
 * @param  {Array***REMOVED*** result
 * @return {Array***REMOVED***
 */
function flattenDown (array, result) {
  for (var i = 0; i < array.length; i++) {
    var value = array[i]

    if (Array.isArray(value)) {
      flattenDown(value, result)
  ***REMOVED*** else {
      result.push(value)
  ***REMOVED***
***REMOVED***

  return result
***REMOVED***

/**
 * Flatten an array with depth.
 *
 * @param  {Array***REMOVED***  array
 * @param  {Array***REMOVED***  result
 * @param  {number***REMOVED*** depth
 * @return {Array***REMOVED***
 */
function flattenDownDepth (array, result, depth) {
  depth--

  for (var i = 0; i < array.length; i++) {
    var value = array[i]

    if (depth > -1 && Array.isArray(value)) {
      flattenDownDepth(value, result, depth)
  ***REMOVED*** else {
      result.push(value)
  ***REMOVED***
***REMOVED***

  return result
***REMOVED***
