// @grpc/grpc-js/build/src/status-builder.js
"use strict";
/*
 * Copyright 2019 gRPC authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
/**
 * A builder for gRPC status objects.
 */
class StatusBuilder {
  ***REMOVED***
        this.code = null;
        this.details = null;
        this.metadata = null;
  ***REMOVED***
    /**
     * Adds a status code to the builder.
     */
    withCode(code) {
        this.code = code;
        return this;
  ***REMOVED***
    /**
     * Adds details to the builder.
     */
    withDetails(details) {
        this.details = details;
        return this;
  ***REMOVED***
    /**
     * Adds metadata to the builder.
     */
    withMetadata(metadata) {
        this.metadata = metadata;
        return this;
  ***REMOVED***
    /**
     * Builds the status object.
     */
    build() {
        const status = {***REMOVED***;
        if (this.code !== null) {
            status.code = this.code;
      ***REMOVED***
        if (this.details !== null) {
            status.details = this.details;
      ***REMOVED***
        if (this.metadata !== null) {
            status.metadata = this.metadata;
      ***REMOVED***
        return status;
  ***REMOVED***
***REMOVED***
exports.StatusBuilder = StatusBuilder;
